src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/Shake128.cs

```diff
@@ -100,12 +110,17 @@ public byte[] GetHashAndReset(int outputLength)
         /// <param name="destination">The buffer to fill with the hash.</param>
         /// <exception cref="ObjectDisposedException">The object has already been disposed.</exception>
         /// <seealso cref="GetCurrentHash(Span{byte})" />
-        public void GetHashAndReset(Span<byte> destination)
+        public unsafe void GetHashAndReset(Span<byte> destination)
         {
             CheckDisposed();
 
-            _hashProvider.Finalize(destination);
-            _hashProvider.Reset();
+            ConcurrentGuard<IntPtr, object?>((IntPtr)(&destination), static (IntPtr destinationPointer, ref LiteXof hashProvider) =>
+            {
+                Span<byte> destination = *(Span<byte>*)destinationPointer;
```

Comment:
I think that something like this will also work, and avoid the address-of-managed-object, and the delegate:


```C#
// ConcurrencyBlock.OpenSsl.cs
internal struct ConcurrencyBlock
{
    private int _count;

    internal static Scope Enter(ref ConcurrencyBlock block)
    {
        int count = Interlocked.Increment(ref block._count);
        
        if (count != 1)
        {
            Interlocked.Decrement(ref block._count);
            throw new Exception();
        }
        
        return new Scope(ref block._count);
    }
   
    internal ref struct Scope
    {
        private ref int _parentCount;
        
        internal Scope(ref int parentCount)
        {
            _parentCount = parentCount;
        }
        
        internal void Dispose()
        {
            Interlocked.Decrement(ref _parentCount);
        }
    }
}
```

```C#
// ConcurrencyBlock.Empty.cs
internal struct ConcurrencyBlock
{
    internal static Scope Enter(ref ConcurrencyBlock block)
    {
        return default;
    }
   
    internal ref struct Scope
    {
        internal void Dispose()
        {
        }
    }
}
```

And then the local changes become


```C#
private ConcurrencyBlock _block;

...

using (ConcurrencyBlock.Enter(ref _block))
{
    _hashProvider.Finalize(destination);
    _hashProvider.Reset();
}
```

According to SharpLab, the JIT sees that Empty doesn't do anything and gets rid of all of it:

[With a body](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgAwAEmuAdACoAWATgKYQAmsA5gNxFGYDMFbGQDCZAN5Ey0sgAcGUAG4QETUQHsYAYwCuDZtoCeAIQA26rQGsyAfWDmrPQjLJSZAsrARkAsgApKchQASjdpSWcXd3I/EU1dfSYjMwtLKgBRGFUGP2YAM1t7VODQyKiJMPL3AHYyFCoAGSSOBDonKrIAX0ruwl6iLyYGGAhTOoQGHS1vOO09Ay0TB0siCJd5JRU1L1stdR0sp0rB4dGKXAA2MgBlPdk1TOzcpgLZhIWl1LIiq1KXNaqOz2B28AF4yABJLJDZZMVhUKFaZhgJIIZ4FH5pGzArLBdpVSouKAFPw47wAQnBuD+VQBHWkUOysPhABEmEimCisujvssqNj9rj8fSyK0GOoAO5kGBMKXpAAeWiYsgQUE0fjxhJkvQ6WukmFqMqlt3U9x5mP5ZM1ZS6YWO0NOY3y40m0xudyYlTpUQ2ylUZGdOxssggBgQcRBwqies8DpGYxNZsDWTkodREdxMe9HWDaayGbBqbDBajLh1BJtRLjZ0wABYyCyoChZOoUEwNVmYy5GTDUnCqGyOVy0c7c8XBQhrR1y9Jep0gA===)  vs  [Empty](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUJgAwAEmuAdACoAWATgKYQAmsA5gNxFGYDMFbGQDCZAN5Ey0sgAcGUAG4QETUQHsYAYwCuDZtoCeAIQA26rQGsyAfWDmrPQjLJSZAsrARkAsgApKchQASjdpSWcXd3I/EU1dfSYjMwtLKgBRGFUGP2YAM1t7VODQyKiJMPL3AHYyFCoAGSSOBDonKrIAX0ruwl6iLyYGGAhTOoQGHS1vOO09Ay0TB0siCJdB4dGKXAA2MgBlLXVZNUzs3KYC2YSFpdSyIqtSlzWqzFrWS4gdUwR2mV6MkqGxGY3y40m0wORxOlVe5RBW0wABYyAARKAoWTqFBMPzPKrwqKAgFETpAA==)

Reply:
That does seem... better.